llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -1848,7 +1850,95 @@ void lowerGlobalsToVector(Module &M) {
 
 } // namespace
 
-PreservedAnalyses SYCLLowerESIMDPass::run(Module &M, ModuleAnalysisManager &) {
+bool SYCLLowerESIMDPass::prepareForAlwaysInliner(Module &M) {
+
+  auto markAlwaysInlined = [](Function &F) -> bool {
+    if (F.hasFnAttribute(llvm::Attribute::NoInline))
+      F.removeFnAttr(llvm::Attribute::NoInline);
+    if (F.hasFnAttribute(llvm::Attribute::InlineHint))
+      F.removeFnAttr(llvm::Attribute::InlineHint);
+    F.addFnAttr(llvm::Attribute::AlwaysInline);
+    return true;
+  };
+
+  bool NeedInline = false;
+  for (auto &F : M) {
+    // If some function already has 'alwaysinline' attribute, then request
+    // inliner pass.
+    if (F.hasFnAttribute(Attribute::AlwaysInline)) {
+      NeedInline = true;
+      continue;
+    }
+
+    // VC BE forbids 'alwaysinline' and "VCStackCall" on the same function.
+    // Such function may be used in other module, we cannot remove it
+    // after inlining.
+    if (F.hasFnAttribute(llvm::genx::VCFunctionMD::VCStackCall))
+      continue;
+
+    if (isESIMDKernel(F))
+      continue;
+
+    if (isSlmAllocatorConstructor(F) || isSlmAllocatorDestructor(F)) {
+      // slm_allocator constructor and destructor must be inlined
+      // to help SLM reservation analysis.
+      NeedInline |= markAlwaysInlined(F);
```

Comment:
`markAlwaysInlined` always returns true, should we make it return void and have this be
`markAlwaysInlined(F)`?

Reply:
The original version of markAlwaysInlined() could return false in some cases, e.g. when F already had always_inline attribute.
Even with markAlwaysInlined always returning true, the current variant looks shorter and it connects two assignments that must be done always together into one.
```
NeedInline |= markAlwaysInlined(F)
```
instead of 
```
markAlwaysInlined(F);
NeedInline = true;
```

More details + bonus question below ))

Originally I thought - if we have a function that is already has `alwaysinline` attribute, then let's call always-inliner to avoid potential problems.

I realize now that potentially we may have a situation when there is only some function F marked as 'alwaysinline' and that doesn't have calls of it because all calls are already inlined. (And inliner doesn't necessary delete functions after inlining their calls). In such cases the invocation of 'AlwaysInlinerPass' is redundant.

We have 2 alternatives here. Which one do you think we should choose?
1) to rely on previous code. i.e assume that if some F has 'alwaysinline', then calls of F are inlined. (return false from markAlwaysInlined()
2) analyze the users of F. If there is some call of F (and F has 'alwaysinline), then force AlwaysInlinerPass. Otherwise, return false from markAlwaysInlined()

Reply:
> foo that is marked alwaysinline. Today if they use -O0 (with or without slm_init) that function will not get inlined ever because of O0

It is not correct, I believe: https://godbolt.org/z/dbYeq6rfn
If user's function is marked as alwaysinline that is the request, not a recommendation or hint. It is honored even with -O0. The inlining is done in FE, before sycl-post-link. In this case there will be a function foo, but no calls of it.

> but we need to call inliner because of slm_init?

Code before the patch inlined everything except 'noinline' and VCStackCall functions claiming GPU RT requires that. It may be an obsolete requirement - I don't know and that is why I added TODO comment: https://github.com/intel/llvm/blob/af09ef469fcebe888fa301bdc05ac08b7648087a/llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp#L1889

Inlining for -O0 is required for slm_init() in the first place, it is also required to handle use-defs having this pattern:
: offset = __esimd_slm_alloc(); /*usage_scope*/; __esimd_slm_free(offset);
```
{
  slm_allocator<size> ExtraMem; // ExtraMem.offset = __esimd_slm_alloc(size)
  ...
  // call of destructor for ExtraMem:  esimd_slm_free(ExtraMem.offset)
```
Without inlining of ~slm_allocator.ctor and ~slm_allocator() it does not seem doable.
Honestly speaking, this patch does not enable slm_allocator for -O0 because the use-def patter is still too difficult for analysis and more changes would be needed, but inlining is required for sure.

Reply:
> It is not correct,
Sorry you're right, I forgot to check for callsites of the function in my test, I forgot the inliner doesn't delete it even if never called.

> Code before the patch inlined everything except 'noinline' and VCStackCall functions claiming GPU RT requires that. 
Ah yeah I see that now